src/main/ncdio_pio.F90.in

```diff
@@ -1693,16 +1695,49 @@ contains
                 call pio_setframe(ncid, vardesc, int(nt,kind=Pio_Offset_Kind))
              end if
 #if ({ITYPE}==TYPELOGICAL)
-             allocate(idata(size(data)))
-             call pio_read_darray(ncid, vardesc, iodesc_plus%iodesc, idata, status)
-             data = (idata == 1)
-             if ( any(idata /= 0 .and. idata /= 1) )then
-                call shr_sys_abort(' ERROR: read in bad integer value(s) for logical data'//errMsg(sourcefile, __LINE__))
-             end if
-             deallocate( idata )
+             if (xtype == PIO_INT .or. xtype == PIO_REAL) then
+                allocate(idata(size(data)))
+                call pio_read_darray(ncid, vardesc, iodesc_plus%iodesc, idata, status)
+                data = (idata == 1)
+                if ( any(idata /= 0 .and. idata /= 1) )then
+                   call shr_sys_abort(' ERROR: read in bad integer value(s) for logical data'//errMsg(sourcefile, __LINE__))
+                end if
+                deallocate( idata )
+             else if (xtype == PIO_DOUBLE) then
+                allocate(ddata(size(data)))
+                call pio_read_darray(ncid, vardesc, iodesc_plus%iodesc, ddata, status)
+                data = (ddata == 1.)
+                if ( any(ddata /= 0. .and. ddata /= 1.) )then
+                   call shr_sys_abort(' ERROR: read in bad value(s) for logical data'//errMsg(sourcefile, __LINE__))
+                end if
+                deallocate( ddata )
+             endif
 #else
              call pio_seterrorhandling(ncid, PIO_BCAST_ERROR)
-             call pio_read_darray(ncid, vardesc, iodesc_plus%iodesc, data, status)
+#if ({ITYPE}==TYPEINT)
+             if (xtype == PIO_INT .or. xtype == PIO_REAL) then
+                call pio_read_darray(ncid, vardesc, iodesc_plus%iodesc, data, status)
```

Comment:
- [ ] So, to confirm: it's always safe to read a single-precision real on the file into an integer in memory?